astronomer/astro-cli
# Feature: Add --no-cache flag to astro deploy and astro dev start

## Description

- Add `--no-cache` flag on `astro dev start` and `astro deploy` commands. This allows users to force building docker images from scratch (ignore docker build cache).
- Had to refactor the ContainerHandler and ImageHandler interfaces and implementations for Docker and Podman inside the `airflow` package in order to accept the NoCache parameter.

## 🎟 Issue(s)

https://github.com/astronomer/issues/issues/3203

## 🧪 Functional Testing

**To test the --no-cache on astro deploy**:
- Create an Astronomer workspace, and `astro workspace switch` to this namespace
- Create an Astronomer deployment
- Use the astro deploy with --no-cache flag, you'll have to deploy at least 2 times, to check that the second time, the image is built again without cache:
```bash
astro deploy <DEPLOYMENT> --no-cache
```

**To test the --no-cache on astro dev start**:
- Initialize a local project with `astro init`
- Start the local project once (this will build the image without cache as this is the first build):
```bash
astro dev start
```
- Start a second time to check that the command uses cache (logs should print `CACHED` when building the image):
```bash
astro dev start
```
- Start a third time with the --no-cache flag (this should build the image from scratch without using docker build cache):
```bash
astro dev start
```

## 📋 Checklist

- [x] Rebased from the main (or release if patching) branch (before testing)
- [x] Ran `make test` before taking out of draft
- [x] Added/updated applicable tests
- [x] Tested against [Houston-API](https://github.com/astronomer/houston-api/) and [Astronomer](https://github.com/astronomer/astronomer/) (if necessary).
- [ ] Communicated to/tagged owners of respective clients potentially impacted by these changes.
- [ ] Updated any related [documentation](https://github.com/astronomer/docs/)
